js/src/common/models/User.ts

```diff
@@ -0,0 +1,165 @@
+  }
+  newNotificationCount() {
+    return Model.attribute<number | undefined>('newNotificationCount').call(this);
+  }
+
+  discussionCount() {
+    return Model.attribute<number | undefined>('discussionCount').call(this);
+  }
+  commentCount() {
+    return Model.attribute<number | undefined>('commentCount').call(this);
+  }
+
+  canEdit() {
+    return Model.attribute<boolean | undefined>('canEdit').call(this);
+  }
+  canEditCredentials() {
+    return Model.attribute<boolean | undefined>('canEditCredentials').call(this);
+  }
+  canEditGroups() {
+    return Model.attribute<boolean | undefined>('canEditGroups').call(this);
+  }
+  canDelete() {
+    return Model.attribute<boolean | undefined>('canDelete').call(this);
+  }
+
+  color() {
+    return computed<string, User>('displayName', 'avatarUrl', 'avatarColor', (displayName, avatarUrl, avatarColor) => {
+      // If we've already calculated and cached the dominant color of the user's
+      // avatar, then we can return that in RGB format. If we haven't, we'll want
+      // to calculate it. Unless the user doesn't have an avatar, in which case
+      // we generate a color from their display name.
+      if (avatarColor) {
+        return `rgb(${(avatarColor as Color).join(', ')})`;
+      } else if (avatarUrl) {
+        this.calculateAvatarColor();
+        return '';
+      }
+
+      return '#' + stringToColor(displayName as string);
+    }).call(this);
+  }
+
+  protected avatarColor: Color | null = null;
+
+  /**
+   * Check whether or not the user has been seen in the last 5 minutes.
+   */
+  isOnline(): boolean {
+    return dayjs().subtract(5, 'minutes').isBefore(this.lastSeenAt());
+  }
+
+  /**
+   * Get the Badge components that apply to this user.
+   */
+  badges(): ItemList<Mithril.Children> {
+    const items = new ItemList<Mithril.Children>();
+    const groups = this.groups();
+
+    if (groups) {
+      groups.forEach((group) => {
+        items.add(`group${group?.id()}`, GroupBadge.component({ group }));
+      });
+    }
+
+    return items;
+  }
+
+  /**
+   * Calculate the dominant color of the user's avatar. The dominant color will
+   * be set to the `avatarColor` property once it has been calculated.
+   */
+  protected calculateAvatarColor() {
+    const image = new Image();
+    const user = this;
+
+    // @ts-expect-error This shouldn't be failing.
+    image.onload = function (this: HTMLImageElement) {
+      try {
+        const colorThief = new ColorThief();
+        user.avatarColor = colorThief.getColor(this);
+      } catch (e) {
+        // Completely white avatars throw errors due to a glitch in color thief
+        // See https://github.com/lokesh/color-thief/issues/40
+        if (e instanceof TypeError) {
+          user.avatarColor = [255, 255, 255];
+        } else {
+          throw e;
+        }
+      }
+      user.freshness = new Date();
+      m.redraw();
+    };
```

Comment:
This seems to prevent TS running wild:

```suggestion
    image.addEventListener('load', function (this: HTMLImageElement) {
      try {
        const colorThief = new ColorThief();
        user.avatarColor = colorThief.getColor(this);
      } catch (e) {
        // Completely white avatars throw errors due to a glitch in color thief
        // See https://github.com/lokesh/color-thief/issues/40
        if (e instanceof TypeError) {
          user.avatarColor = [255, 255, 255];
        } else {
          throw e;
        }
      }
      user.freshness = new Date();
      m.redraw();
    });
```